src/displayBackend/common/ConnectorBase.cpp

```diff
@@ -0,0 +1,347 @@
+	uint32_t xfront = xres * 25 / 100;
+	uint32_t xsync  = xres *  3 / 100;
+	uint32_t xblank = xres * 35 / 100;
+
+	uint32_t yfront = yres *  5 / 1000;
+	uint32_t ysync  = yres *  5 / 1000;
+	uint32_t yblank = yres * 35 / 1000;
+
+	uint32_t clock = EDID_REFRESH_RATE_HZ * (xres + xblank) * (yres + yblank);
+
+	/* 10 KHz granularity, little endian. */
+	desc->pixel_clock = toLittleEndian32(clock / 10000);
+
+	pixelData->hactive_lo = xres & 0xff;
+	pixelData->hblank_lo = xblank & 0xff;
+	pixelData->hactive_hblank_hi = (((xres   & 0xf00) >> 4) |
+									((xblank & 0xf00) >> 8));
+
+	pixelData->vactive_lo = yres & 0xff;
+	pixelData->vblank_lo = yblank & 0xff;
+	pixelData->vactive_vblank_hi = (((yres   & 0xf00) >> 4) |
+									((yblank & 0xf00) >> 8));
+
+	pixelData->hsync_offset_lo = xfront & 0xff;
+	pixelData->hsync_pulse_width_lo = xsync & 0xff;
+
+	pixelData->vsync_offset_pulse_width_lo = (((yfront & 0x00f) << 4) |
+											  ((ysync  & 0x00f) << 0));
+	pixelData->hsync_vsync_offset_pulse_width_hi = (((xfront & 0x300) >> 2) |
+													((xsync  & 0x300) >> 4) |
+													((yfront & 0x030) >> 2) |
+													((ysync  & 0x030) >> 4));
+
+	pixelData->width_mm_lo = xmm & 0xff;
+	pixelData->height_mm_lo = ymm & 0xff;
+	pixelData->width_height_mm_hi = (((xmm & 0xf00) >> 4) |
+									 ((ymm & 0xf00) >> 8));
+
+	pixelData->hborder = 0;
+	pixelData->vborder = 0;
+
+	pixelData->misc = DRM_EDID_PT_SEPARATE_SYNC;
+}
+
+void ConnectorBase::edidPutTimings(edid* edidBlock)
+{
+	/*
+	 * Established timing bitmap. Supported bitmap for (formerly)
+	 * very common timing modes: we do not want to provide any.
+	 */
+	memset(&edidBlock->established_timings, 0,
+		   sizeof(edidBlock->established_timings));
+
+	/*
+	 * Standard timing information. Up to 8 2-byte fields describing
+	 * standard display modes. Unused fields are filled with 01.
+	 *
+	 * We do not provide any, but detailed timings.
+	 */
+	memset(edidBlock->standard_timings, 0x01,
+		   sizeof(edidBlock->standard_timings));
+
+	/*
+	 * Detailed timing descriptors, in decreasing preference order,
+	 * followed by Display descriptors.
+	 *
+	 * We only provide a single timing here which corresponds to
+	 * XenStore configuration of this connector.
+	 */
+	edidPutDetailedTiming(edidBlock, 0, mCfgWidth, mCfgHeight, EDID_DPI);
+}
+
+size_t ConnectorBase::getEDID(grant_ref_t startDirectory, uint32_t size)
+{
+	GrantRefs refs;
+
+	pgDirGetBufferRefs(mDomId, startDirectory, size, refs);
+	if (!refs.size())
+	{
+		throw Exception("Cannot get grant references of the EDID buffer",
+						EINVAL);
+	}
+
+	std::unique_ptr<XenGnttabBuffer> edidBuffer(
+		new XenGnttabBuffer(mDomId, refs.data(), refs.size()));
+
+	if (!edidBuffer)
+	{
+		throw Exception("Failed to create shared buffer for EDID", ENOMEM);
+	}
+
+	memset(edidBuffer->get(), 0, XENDISPL_EDID_BLOCK_SIZE);
+
+	auto edidBlock = static_cast<edid*>(edidBuffer->get());
+
+	edidPutEssentials(edidBlock);
```

Comment:
what is the sense of unique_ptr here ?
use std::shared_ptr here and in the method declarations